### ISSUE TITLE
Getting Cannot read property 'call' of undefined when building in production mode

### ISSUE THREAD

**Issue by stubbsy345** [akveo/nebular#922]
**I'm submitting a ...**  (check one with "x")

* [X] bug report
* [ ] feature request

### Issue description

Getting the following error when trying to deploy in --prod mode:

```
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'call' of undefined
TypeError: Cannot read property 'call' of undefined
    at e.isRedirectResult (main.7829c0b210d0c660fe04.js:1)
    at e.authenticate (main.7829c0b210d0c660fe04.js:1)
    at t.authenticate (main.7829c0b210d0c660fe04.js:1)
    at e._tapNext (main.7829c0b210d0c660fe04.js:1)
    at e._next (main.7829c0b210d0c660fe04.js:1)
    at e.next (main.7829c0b210d0c660fe04.js:1)
    at e._next (main.7829c0b210d0c660fe04.js:1)
    at e.next (main.7829c0b210d0c660fe04.js:1)
    at t._subscribe (main.7829c0b210d0c660fe04.js:1)
    at t._trySubscribe (main.7829c0b210d0c660fe04.js:1)
    at e.isRedirectResult (main.7829c0b210d0c660fe04.js:1)
    at e.authenticate (main.7829c0b210d0c660fe04.js:1)
    at t.authenticate (main.7829c0b210d0c660fe04.js:1)
    at e._tapNext (main.7829c0b210d0c660fe04.js:1)
    at e._next (main.7829c0b210d0c660fe04.js:1)
    at e.next (main.7829c0b210d0c660fe04.js:1)
    at e._next (main.7829c0b210d0c660fe04.js:1)
    at e.next (main.7829c0b210d0c660fe04.js:1)
    at t._subscribe (main.7829c0b210d0c660fe04.js:1)
    at t._trySubscribe (main.7829c0b210d0c660fe04.js:1)
    at O (polyfills.e3493840eec9391c0c02.js:1)
    at O (polyfills.e3493840eec9391c0c02.js:1)
    at polyfills.e3493840eec9391c0c02.js:1
    at t.invokeTask (polyfills.e3493840eec9391c0c02.js:1)
    at Object.onInvokeTask (main.7829c0b210d0c660fe04.js:1)
    at t.invokeTask (polyfills.e3493840eec9391c0c02.js:1)
    at e.runTask (polyfills.e3493840eec9391c0c02.js:1)
    at v (polyfills.e3493840eec9391c0c02.js:1)
```

Have spent a lot of time trying to find a solution but nothing appears to work. Currently on 2.0.0rc10. Was wondering if this was an issue you had encountered before. Only happens in --prod build.

**Comment by nnixaa** [comment]
Hi @stubbsy345, is there a chance of providing a reproducible example? You can push it to stackblitz, for example.

**Comment by stubbsy345** [comment]
I am using this exact overwrite code that you suggested in a thread to get the authorisation working with azure b2c. Works well when build in dev mode, but goes to original error in prod build.

(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token'

This does not seem to work in production mode however, there is some problem with overwriting the enum value in —prod.

**Comment by stubbsy345** [comment]
My auth code is exactly the same as @jjgriff used that you both developed in associated issues, I spoke with @jjgriff and he is encountering the same problem in production mode.

**Comment by stubbsy345** [comment]
Original correction that fixes dev build but then fails in prod build:

```(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token'```

```
import { ModuleWithProviders, NgModule, Optional, SkipSelf, Injectable, Inject } from '@angular/core';
import { CommonModule } from '@angular/common';
// tslint:disable-next-line:max-line-length
import { NbOAuth2AuthStrategy, NbAuthModule, NbOAuth2ResponseType, NbAuthOAuth2JWTToken, NbOAuth2AuthStrategyOptions, NbAuthStrategyClass } from '@nebular/auth';
import { NbSecurityModule, NbRoleProvider } from '@nebular/security';
import { of } from 'rxjs/observable/of';

import { throwIfAlreadyLoaded } from './module-import-guard';
import { DataModule } from './data/data.module';
import { AppMonitoringService } from './monitoring/app-monitoring.service';
import { map } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { ActivatedRoute } from '@angular/router';
import { NB_WINDOW } from '@nebular/theme';

const socialLinks = [
  {
    url: 'https://www.facebook.com/pixeldoctor/',
    target: '_blank',
    icon: 'socicon-facebook',
  },
];

export class NbSimpleRoleProvider extends NbRoleProvider {
  getRole() {
    // here you could provide any role based on any auth flow
    return of('guest');
  }
}

// Override 'token' to 'id_token' for Azure AD B2C
(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token';

// Create new token for Azure auth so it returns id_token instead of access_token
export class NbAuthAzureToken extends NbAuthOAuth2JWTToken {

  // let's rename it to exclude name clashes
  static NAME = 'nb:auth:azure:token';

  getValue(): string {
    return this.token.id_token;
  }
}

@Injectable()
export class NbAzureADB2CAuthStrategy extends NbOAuth2AuthStrategy {

  // we need this methos for strategy setup
  static setup(options: NbOAuth2AuthStrategyOptions): [NbAuthStrategyClass, NbOAuth2AuthStrategyOptions] {
    return [NbAzureADB2CAuthStrategy, options];
  }

  protected redirectResults: any = {
    [NbOAuth2ResponseType.CODE]: () => of(null),

    [NbOAuth2ResponseType.TOKEN]: () => {

      // queryParams here as token returned not as hash
      return of(this.activatedRoute.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => !!(params && (params.id_token || params.error))),
      );
    },
  };

  constructor(http: HttpClient,
              private activatedRoute: ActivatedRoute,
              @Inject(NB_WINDOW) window: any) {
    super(http, activatedRoute, window);
  }
}

export const NB_CORE_PROVIDERS = [
  ...DataModule.forRoot().providers,
  ...NbAuthModule.forRoot({
    strategies: [
      NbAzureADB2CAuthStrategy.setup({
        name: 'AzureADB2C',
        clientId: '*************************',
        clientSecret: '',
        authorize: {
          // tslint:disable-next-line:max-line-length
          endpoint: 'https://login.microsoftonline.com/tfp/***********/b2c_1_pxdrsignin/oauth2/v2.0/authorize',
          responseType: NbOAuth2ResponseType.TOKEN, // return id_token instead of token for Azure auth
          scope: '***********/api/profile openid profile',
          redirectUri: 'https://localhost:5001/auth/callback',
        },
        token: {
          class: NbAuthAzureToken,
        },
        redirect: {
          success: '/dashboard/overview',
        },
      }),
    ],
    forms: {
      login: {
        socialLinks: socialLinks,
      },
      register: {
        socialLinks: socialLinks,
      },
    },
  }).providers,

  NbSecurityModule.forRoot({
    accessControl: {
      guest: {
        view: '*',
      },
      user: {
        parent: 'guest',
        create: '*',
        edit: '*',
        remove: '*',
      },
    },
  }).providers,
  {
    provide: NbRoleProvider, useClass: NbSimpleRoleProvider,
  },
  AppMonitoringService,
];

@NgModule({
  imports: [
    CommonModule,
  ],
  exports: [
    NbAuthModule,
  ],
  declarations: [],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }

  static forRoot(): ModuleWithProviders {
    return <ModuleWithProviders>{
      ngModule: CoreModule,
      providers: [
        ...NB_CORE_PROVIDERS,
        NbAzureADB2CAuthStrategy,
      ],
    };
  }
}
```

**Comment by stubbsy345** [comment]
This is the core module. For some reason in dev mode this code: 
```(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token'``` 
fixes the issue, but not in prod build. I know at @jjgriff93 is suffering from the same problem.

```
import { ModuleWithProviders, NgModule, Optional, SkipSelf, Injectable, Inject } from '@angular/core';
import { CommonModule } from '@angular/common';
// tslint:disable-next-line:max-line-length
import { NbOAuth2AuthStrategy, NbAuthModule, NbOAuth2ResponseType, NbAuthOAuth2JWTToken, NbOAuth2AuthStrategyOptions, NbAuthStrategyClass } from '@nebular/auth';
import { NbSecurityModule, NbRoleProvider } from '@nebular/security';
import { of } from 'rxjs/observable/of';

import { throwIfAlreadyLoaded } from './module-import-guard';
import { DataModule } from './data/data.module';
import { AppMonitoringService } from './monitoring/app-monitoring.service';
import { map } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { ActivatedRoute } from '@angular/router';
import { NB_WINDOW } from '@nebular/theme';

const socialLinks = [
  {
    url: 'https://www.facebook.com/pixeldoctor/',
    target: '_blank',
    icon: 'socicon-facebook',
  },
];

export class NbSimpleRoleProvider extends NbRoleProvider {
  getRole() {
    // here you could provide any role based on any auth flow
    return of('guest');
  }
}

// Override 'token' to 'id_token' for Azure AD B2C
(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token';

// Create new token for Azure auth so it returns id_token instead of access_token
export class NbAuthAzureToken extends NbAuthOAuth2JWTToken {

  // let's rename it to exclude name clashes
  static NAME = 'nb:auth:azure:token';

  getValue(): string {
    return this.token.id_token;
  }
}


@Injectable()
export class NbAzureADB2CAuthStrategy extends NbOAuth2AuthStrategy {

  // we need this methos for strategy setup
  static setup(options: NbOAuth2AuthStrategyOptions): [NbAuthStrategyClass, NbOAuth2AuthStrategyOptions] {
    return [NbAzureADB2CAuthStrategy, options];
  }

  protected redirectResults: any = {
    [NbOAuth2ResponseType.CODE]: () => of(null),

    [NbOAuth2ResponseType.TOKEN]: () => {

      // queryParams here as token returned not as hash
      return of(this.activatedRoute.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => !!(params && (params.id_token || params.error))),
      );
    },
  };

  constructor(http: HttpClient,
              private activatedRoute: ActivatedRoute,
              @Inject(NB_WINDOW) window: any) {
    super(http, activatedRoute, window);
  }
}

export const NB_CORE_PROVIDERS = [
  ...DataModule.forRoot().providers,
  ...NbAuthModule.forRoot({
    strategies: [
      NbAzureADB2CAuthStrategy.setup({
        name: 'AzureADB2C',
        clientId: '7f1d19f9-80ed-43c8-86df-ed78b2bd6275',
        clientSecret: '',
        authorize: {
          endpoint: 'https://PixelDr.b2clogin.com/pixeldr.onmicrosoft.com/oauth2/v2.0/authorize',
          responseType: NbOAuth2ResponseType.TOKEN, // return id_token instead of token for Azure auth
          scope: 'https://PixelDr.onmicrosoft.com/api/profile openid profile',
          redirectUri: 'https://localhost:5001/auth/callback',
          params: {
            'p': 'B2C_1A_signuporsignin'// Select the AADB2C policy to use
          }
        },
        token: {
          class: NbAuthAzureToken,
        },
        redirect: {
          success: '/dashboard/overview',
        },
      }),
    ],
    forms: {
      login: {
        socialLinks: socialLinks,
      },
      register: {
        socialLinks: socialLinks,
      },
    },
  }).providers,

  NbSecurityModule.forRoot({
    accessControl: {
      guest: {
        view: '*',
      },
      user: {
        parent: 'guest',
        create: '*',
        edit: '*',
        remove: '*',
      },
    },
  }).providers,
  {
    provide: NbRoleProvider, useClass: NbSimpleRoleProvider,
  },
  AppMonitoringService,
];

@NgModule({
  imports: [
    CommonModule,
  ],
  exports: [
    NbAuthModule,
  ],
  declarations: [],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }

  static forRoot(): ModuleWithProviders {
    return <ModuleWithProviders>{
      ngModule: CoreModule,
      providers: [
        ...NB_CORE_PROVIDERS,
        NbAzureADB2CAuthStrategy,
      ],
    };
  }
}
```

**Comment by dpollastrini** [comment]
I bumped into this too while using 'id_token token' in the response type setup rather than an enum value. It seems the "call" in question is to an object named based on the enum values specified NbOAuth2ResponseType (i.e., TOKEN _**or**_ CODE). There is apparently on "ID_TOKEN" option hence the error indicating no "call" on undefined. I have a case where my response type needs to be "id_token token", a combination (or at least an "id_token" request, validate id, then "token" request (aka access token), which is a suggested oauth2 implicit flow pattern).  This doesn't seem possible with how the response types are "hard coded" in the NbAuth architecture.  ID_TOKEN isn't an option and only one response type can be specified at a time (providers like IdentityServer4 support multiple response types in a single request). Requesting scopes like profile or openid, require (should) an ID_TOKEN response type. My guess is that handling multiple tokens in the call back would require some additional changes beyond just enabling different response type combinations.

**Comment by stubbsy345** [comment]
@dpollastrini thanks for your response. Did you find a work around? We are hoping to begin testing of our tool with users but not being able to build in prod mode is causing a significant problem.

**Comment by dpollastrini** [comment]
@stubbsy345, the overrides you're using work for me; however, in my case, for various reasons, I actually need to extend the response types, not override the existing. While adding something like ['id_token']: () => .... to redirectResults will get you authenticated, you run into post-authentication issues later in the flow (e.g, error: "'id_token' responseType is not supported, only 'token' and 'code' are supported now'"). There are likely additional overrides and/or hacks that can workaround this type of error, but I've chosen not to scuttle down this particular rabbit hole right now.

For my case, I'm opting for a different 3rd-party module that provides flexible OAuth2 services...at least until the NbAuth module matures a bit. I'll still be importing the NbSecurity module, which is awesome sauce (with a couple of gotchas hooking the role provider in to auth guards). 

These Nebular folks are excellent and I'm sure the auth module will eventually evolve into something more flexible.

**Comment by stubbsy345** [comment]
Yes indeed they work for me when built in dev, however, when built in production mode the override doesn’t seem to work. I have tried to change some of the production build settings but with no success. Do you have any idea why in production the overwrite would no longer work?

**Comment by nnixaa** [comment]
Hey Guys, take a look at the following comment https://github.com/akveo/nebular/issues/541#issuecomment-411776758, I believe it could be useful.

**Comment by jjgriff93** [comment]
Hi @nnixaa I've done some investigation and deployed the code we put together into an Azure live environment, and when hitting the auth-guard and beginning the redirection to sign in, it hits the error:

```
at e.isRedirectResult (main.180b46e5a8b48abd0f58.js:formatted:46543)
    at e.authenticate (main.180b46e5a8b48abd0f58.js:formatted:46497)
    at t.authenticate (main.180b46e5a8b48abd0f58.js:formatted:26139)
    at t.logIn (main.180b46e5a8b48abd0f58.js:formatted:50087)
    at Object.handleEvent (main.180b46e5a8b48abd0f58.js:formatted:50116)
    at Object.handleEvent (main.180b46e5a8b48abd0f58.js:formatted:16549)
    at Object.handleEvent (main.180b46e5a8b48abd0f58.js:formatted:17125)
    at Ko (main.180b46e5a8b48abd0f58.js:formatted:14891)
    at main.180b46e5a8b48abd0f58.js:formatted:15237
```

This is the line where it's happening in the nebular auth code (`oauth2-strategy.js`, line 208):
```
e.prototype.isRedirectResult = function() {
                return this.redirectResults[this.responseType].call(this)
            }
```

When debugging and hitting the breakpoint, this is the call stack:

- this: e
-- activatedRoute: t {url: e, params: e, queryParams: e, fragment: e, data: e, …}
-- clientAuthMethod: (...)
-- defaultOptions: e
-- authorize: {endpoint: "authorize", responseType: "code"}
-- baseEndpoint: ""
-- clientAuthMethod: "none"
-- clientId: ""
-- clientSecret: ""
-- defaultErrors: ["Something went wrong, please try again."]
-- defaultMessages: ["You have been successfully authenticated."]
-- redirect: {success: "/", failure: null}
-- refresh: {endpoint: "token", grantType: "refresh_token"}
-- token: {endpoint: "token", grantType: "authorization_code", requireValidToken: false, class: ƒ}
-- __proto__: r
-- http: t {handler: t}
-- options:
-- authorize: {endpoint: "https://XXXXX.b2clogin.com/XXXXX.onmicrosoft.com/oauth2/v2.0/authorize", responseType: "token", scope: "https://XXXXX.onmicrosoft.com/api/profile openid profile", redirectUri: "nullauth/callback", params: {…}}
-- baseEndpoint: ""
-- clientAuthMethod: "none"
-- clientId: "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
-- clientSecret: ""
-- defaultErrors: ["Something went wrong, please try again."]
-- defaultMessages: ["You have been successfully authenticated."]
-- name: "AzureADB2C"
-- redirect: {success: "/dashboard/overview", failure: null}
-- refresh: {endpoint: "token", grantType: "refresh_token"}
-- token: {endpoint: "token", grantType: "authorization_code", requireValidToken: false, class: ƒ}
-- __proto__: Object
-- redirectResultHandlers: {code: ƒ, id_token: ƒ}
-- redirectResults: {code: ƒ, id_token: ƒ}
-- responseType: (...)
-- route: t {url: e, params: e, queryParams: e, fragment: e, data: e, …}
-- window: Window {postMessage: ƒ, blur: ƒ, focus: ƒ, close: ƒ, frames: Window, …}
-- __proto__: e
- this: e

It seems that the overrides we applied Dmitry when originally getting Azure ADB2C to work aren't being applied for some reason in live. This is the output I have at the exact same breakpoint when in dev mode locally:

- this: NbAzureADB2CAuthStrategy
-- activatedRoute: ActivatedRoute {url: BehaviorSubject, params: BehaviorSubject, queryParams: BehaviorSubject, fragment: BehaviorSubject, data: BehaviorSubject, …}
-- clientAuthMethod: (...)
-- defaultOptions: NbOAuth2AuthStrategyOptions {baseEndpoint: "", clientId: "", clientSecret: "", clientAuthMethod: "none", redirect: {…}, …}
-- http: HttpClient {handler: HttpInterceptingHandler}
-- options: {baseEndpoint: "", clientId: "XXXXXXXXXXXXXXXXXXXXXXXXXXX", clientSecret: "", clientAuthMethod: "none", redirect: {…}, …}
-- redirectResultHandlers: {code: ƒ, id_token: ƒ}
-- redirectResults: {code: ƒ, id_token: ƒ}
-- responseType: (...)
-- route: ActivatedRoute {url: BehaviorSubject, params: BehaviorSubject, queryParams: BehaviorSubject, fragment: BehaviorSubject, data: BehaviorSubject, …}
-- window: Window {postMessage: ƒ, blur: ƒ, focus: ƒ, close: ƒ, frames: Window, …}
-- __proto__: NbOAuth2AuthStrategy

Also one major difference here I'd point out from the error on my thread is this is now happening when initiating the call to authenticate, instead of on the callback. Unlike before it's not even directing me to Azure B2C, this error is being returned beforehand.

Any ideas why this would be behaving differently in Angular --prod mode and how we could fix this?

Thanks as always for your help, we're very close to live now, it's just this annoying bug in the way!

**Comment by stubbsy345** [comment]
@nnixaa I can confirm that the debugging output @jjgriff93 is experiencing is the same as we are. Any ideas? I am presuming that production options such as AOT or optimization, which are used during the prod settings, may be affecting the auth directive?

**Comment by jjgriff93** [comment]
Hi @stubbsy345 good find - tried with AoT disabled and it works just fine. I think it must be to do with when the modules are being loaded or the order - on all the posts I've read 'call is undefined' is typically related to a script not being loaded before it's being requested. @nnixaa any ideas?

**Comment by jjgriff93** [comment]
@nnixaa / Akveo team - are you able to help us with this? We're very close to live and using the JIT compiler instead of AoT is slowing down the load times significantly, really appreciate any help!

**Comment by nnixaa** [comment]
Hey Guys, I believe as a workaround we can try to no touche the enum and just rewrite two handler for the `id_token` like this:

```
@Injectable()
export class NbAzureADB2CAuthStrategy extends NbOAuth2AuthStrategy {

  // we need this methos for strategy setup
  static setup(options: NbOAuth2AuthStrategyOptions): [NbAuthStrategyClass, NbOAuth2AuthStrategyOptions] {
    return [NbAzureADB2CAuthStrategy, options];
  }

  protected redirectResultHandlers = {
    [NbOAuth2ResponseType.CODE]: () => {
      return observableOf(this.route.snapshot.queryParams).pipe(
        switchMap((params: any) => {
          if (params.code) {
            return this.requestToken(params.code)
          }

          return observableOf(
            new NbAuthResult(
              false,
              params,
              this.getOption('redirect.failure'),
              this.getOption('defaultErrors'),
              [],
            ));
        }),
      );
    },
    id_token: () => { // <--- !!! HERE
      const module = 'authorize';
      const requireValidToken = this.getOption(`${module}.requireValidToken`);
      return observableOf(this.route.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => {
          if (!params.error) {
            return new NbAuthResult(
              true,
              params,
              this.getOption('redirect.success'),
              [],
              this.getOption('defaultMessages'),
              this.createToken(params, requireValidToken));
          }
          return new NbAuthResult(
            false,
            params,
            this.getOption('redirect.failure'),
            this.getOption('defaultErrors'),
            [],
          );
        }),
        catchError(err => {
          const errors = [];
          if (err instanceof NbAuthIllegalTokenError) {
            errors.push(err.message)
          } else {
            errors.push('Something went wrong.');
          }
          return observableOf(
            new NbAuthResult(
              false,
              err,
              this.getOption('redirect.failure'),
              errors,
            ));
        }),
      );
    },
  };


  protected redirectResults: any = {
    [NbOAuth2ResponseType.CODE]: () => observableOf(null),

    id_token: () => { // <--- !!! HERE
      return observableOf(this.activatedRoute.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => !!(params && (params.id_token || params.error))),
      );
    },
  };

  constructor(http: HttpClient,
              private activatedRoute: ActivatedRoute,
              @Inject(NB_WINDOW) window: any) {
    super(http, activatedRoute, window);
  }
}
```

Let me know if this is working on your side.

**Comment by nnixaa** [comment]
@jjgriff93, @stubbsy345 ping.

**Comment by jjgriff93** [comment]
Hi @nnixaa , apologies for the delayed reply, thank you for putting this code together - I've tried implementing it to test however I'm getting a couple of validation errors:

- On `redirectResultHandlers` I'm getting `[ts]
Property 'redirectResultHandlers' in type 'NbAzureADB2CAuthStrategy' is not assignable to the same property in base type 'NbOAuth2AuthStrategy'.
  Type '{ [NbOAuth2ResponseType.CODE]: () => Observable<NbAuthResult>; id_token: () => Observable<NbAuthResult>; }' is not assignable to type '{ [NbOAuth2ResponseType.CODE]: () => Observable<NbAuthResult>; [NbOAuth2ResponseType.TOKEN]: () => Observable<NbAuthResult>; }'.
    Property '[NbOAuth2ResponseType.TOKEN]' is missing in type '{ [NbOAuth2ResponseType.CODE]: () => Observable<NbAuthResult>; id_token: () => Observable<NbAuthResult>; }'. [2416]
(property) NbAzureADB2CAuthStrategy.redirectResultHandlers: {
    [NbOAuth2ResponseType.CODE]: () => Observable<NbAuthResult>;
    id_token: () => Observable<NbAuthResult>;
}`

- On `route` in both instances of `this.route.snapshot.queryParams` I'm getting `[ts] Property 'route' is private and only accessible within class 'NbOAuth2AuthStrategy'. [2341]
(property) NbOAuth2AuthStrategy.route: any`

The full code is looking like this at the moment:

```
import { ModuleWithProviders, NgModule, Optional, SkipSelf, Injectable, Inject } from '@angular/core';
import { CommonModule } from '@angular/common';
// tslint:disable-next-line:max-line-length
import { NbOAuth2AuthStrategy, NbAuthModule, NbOAuth2ResponseType, NbAuthOAuth2JWTToken, NbOAuth2AuthStrategyOptions, NbAuthStrategyClass, NbAuthResult, NbAuthIllegalTokenError } from '@nebular/auth';
import { NbSecurityModule, NbRoleProvider } from '@nebular/security';
import { of } from 'rxjs/observable/of';

import { throwIfAlreadyLoaded } from './module-import-guard';
import { DataModule } from './data/data.module';
import { AppMonitoringService } from './monitoring/app-monitoring.service';
import { map, switchMap, catchError } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { ActivatedRoute } from '@angular/router';
import { NB_WINDOW } from '@nebular/theme';

const socialLinks = [
  {
    url: 'https://www.facebook.com/pixeldoctor/',
    target: '_blank',
    icon: 'socicon-facebook',
  },
];

export class NbSimpleRoleProvider extends NbRoleProvider {
  getRole() {
    // here you could provide any role based on any auth flow
    return of('guest');
  }
}

// Override 'token' to 'id_token' for Azure AD B2C
(NbOAuth2ResponseType as any)['TOKEN'] = 'id_token';

// Create new token for Azure auth so it returns id_token instead of access_token
export class NbAuthAzureToken extends NbAuthOAuth2JWTToken {

  // let's rename it to exclude name clashes
  static NAME = 'nb:auth:azure:token';

  getValue(): string {
    return this.token.id_token;
  }
}


@Injectable()
export class NbAzureADB2CAuthStrategy extends NbOAuth2AuthStrategy {

  // we need this methos for strategy setup
  static setup(options: NbOAuth2AuthStrategyOptions): [NbAuthStrategyClass, NbOAuth2AuthStrategyOptions] {
    return [NbAzureADB2CAuthStrategy, options];
  }

  protected redirectResultHandlers = {
    [NbOAuth2ResponseType.CODE]: () => {
      return of(this.route.snapshot.queryParams).pipe(
        switchMap((params: any) => {
          if (params.code) {
            return this.requestToken(params.code);
          }

          return of(
            new NbAuthResult(
              false,
              params,
              this.getOption('redirect.failure'),
              this.getOption('defaultErrors'),
              [],
            ));
        }),
      );
    },
    id_token: () => { // <--- !!! HERE
      const module = 'authorize';
      const requireValidToken = this.getOption(`${module}.requireValidToken`);
      return of(this.route.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => {
          if (!params.error) {
            return new NbAuthResult(
              true,
              params,
              this.getOption('redirect.success'),
              [],
              this.getOption('defaultMessages'),
              this.createToken(params, requireValidToken));
          }
          return new NbAuthResult(
            false,
            params,
            this.getOption('redirect.failure'),
            this.getOption('defaultErrors'),
            [],
          );
        }),
        catchError(err => {
          const errors = [];
          if (err instanceof NbAuthIllegalTokenError) {
            errors.push(err.message);
          } else {
            errors.push('Something went wrong.');
          }
          return of(
            new NbAuthResult(
              false,
              err,
              this.getOption('redirect.failure'),
              errors,
            ));
        }),
      );
    },
  };


  protected redirectResults: any = {
    [NbOAuth2ResponseType.CODE]: () => of(null),

    id_token: () => { // <--- !!! HERE
      return of(this.activatedRoute.snapshot.fragment).pipe(
        map(fragment => this.parseHashAsQueryParams(fragment)),
        map((params: any) => !!(params && (params.id_token || params.error))),
      );
    },
  };

  constructor(http: HttpClient,
              private activatedRoute: ActivatedRoute,
              @Inject(NB_WINDOW) window: any) {
    super(http, activatedRoute, window);
  }
}

export const NB_CORE_PROVIDERS = [
  ...DataModule.forRoot().providers,
  ...NbAuthModule.forRoot({
    strategies: [
      NbAzureADB2CAuthStrategy.setup({
        name: 'AzureADB2C',
        clientId: 'XXXXXXXXXXXXXXXXXXXXXXXXXX',
        clientSecret: '',
        authorize: {
          endpoint: 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
          responseType: NbOAuth2ResponseType.TOKEN, // return id_token instead of token for Azure auth
          scope: 'XXXXXXXXXXXXXXXXXXX',
          redirectUri: document.getElementsByTagName('base')[0].href + 'auth/callback',
          params: {
            'p': 'XXXXXXXX'// Select the AADB2C policy to use
          }
        },
        token: {
          class: NbAuthAzureToken,
        },
        redirect: {
          success: '/dashboard/overview',
        },
      }),
    ],
    forms: {
      login: {
        socialLinks: socialLinks,
      },
      register: {
        socialLinks: socialLinks,
      },
    },
  }).providers,

  NbSecurityModule.forRoot({
    accessControl: {
      guest: {
        view: '*',
      },
      user: {
        parent: 'guest',
        create: '*',
        edit: '*',
        remove: '*',
      },
    },
  }).providers,
  {
    provide: NbRoleProvider, useClass: NbSimpleRoleProvider,
  },
  AppMonitoringService,
];

@NgModule({
  imports: [
    CommonModule,
  ],
  exports: [
    NbAuthModule,
  ],
  declarations: [],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }

  static forRoot(): ModuleWithProviders {
    return <ModuleWithProviders>{
      ngModule: CoreModule,
      providers: [
        ...NB_CORE_PROVIDERS,
        NbAzureADB2CAuthStrategy,
      ],
    };
  }
}
```

Any thoughts on what could resolve these errors? Thanks as always for your help

**Comment by jjgriff93** [comment]
Hi @nnixaa - any ideas on the above?

**Comment by nnixaa** [comment]
Hey @jjgriff93, we are releasing 3.0.1 with a couple slight fixes to make the inheritance simpler.
We also added an example of Azure provider, you can find the code either in the PR here https://github.com/akveo/nebular/pull/1065/files#diff-aefaa618a30ec554144a3413976c071c or later on in the docs.
The example will be deployed to our docs so that you can check.
Hope it resolves all the issues!

**Comment by jjgriff93** [comment]
@nnixaa Apologies for the delay - I have implemented 3.0.1 and followed the Azure example and it's working perfectly thank you :)
